### DISCUSSION TITLE
TypeScript generation for internal links

### DISCUSSION THREAD

**Discussion by tomhicks** [vercel/next.js#15402] [Ideas]
Wondering if there's any reason *not* to do something like this to help with internal linking to help create href, as props:
- recursively list all files in /pages
- extract parameter names from inside square brackets (e.g. /posts/[userId]/[id] extracts `userId, id`)
- either apply type restrictions to `<Link />` or make a `getLinkProps(path, params)` function with restricted types

Could also do a slightly different API on `Link`, like:
```
<Link toInternal={{
  route: "/posts/[userId]/[id]",       // this line can only be a known internal page, and...
  params: {id: "abc", userId: "def"}   // enforces the params that must be passed in here
}}/>
```

Someone must have done this already?

If not, I'm willing to do it if it fits with the way routing works (I think it does but I'm fairly new to Next)

**Comment by tomhicks** [comment]
This is the latest commit (not from me!) on our NextJS website:

> Fix mismatch of as and href on listing card

:)

**Comment by tomhicks** [comment]
Here's a not-very-generic, not-very-well-tested implementation of this:

https://gist.github.com/tomhicks/56b6a159ab5e727e5f43353172a49b6d

**Comment by solufa** [comment]
Pathpida may be the solution.

```tsx
import Link from 'next/link'
import { pagesPath } from '../lib/$path' // Automatically generated by pathpida

export default () => {
  const onClick = useCallback(() => {
    router.push(pagesPath.post._pid(1).$url())
  }, [])

  return <>
    <Link href={pagesPath.post._slug(['a', 'b', 'c']).$url()} />
    <div onClick={onClick} />
  </>
}
```

https://github.com/aspida/pathpida

Our Japanese team has been maintaining it for over a year and can answer your questions!
We've submitted it to ProductHunt, and I hope you'll vote for it!

https://www.producthunt.com/posts/pathpida
